dipy/core/gradients.py

```diff
@@ -1343,3 +1343,133 @@ def _check_ornt(ornt):
+    """Extract a set of b0 volumes from a dwi dataset.
+
+    Parameters
+    ----------
+    dwi: ndarray
+        4D diffusion-weighted imaging data (X, Y, Z, N), where N is the number of
+        volumes.
+    b0_mask: ndarray of bool
+        Mask over the time dimension (4th) identifying b0 volumes.
+    group_contiguous_b0: bool, optional
+        If True, each contiguous b0 volumes are grouped together.
+    strategy: str, optional
+        The extraction strategy, of either:
+
+            - first: select the first b0 found.
+            - all: select them all.
+            - mean: average them.
+
+        When used in conjunction with the batch parameter set to True, the strategy
+        is applied individually on each continuous set found.
+
+    Returns
+    -------
+    b0_data : ndarray
+        Extracted b0 volumes.
+    """
+    if not isinstance(b0_mask, np.ndarray) or b0_mask.ndim != 1:
+        raise ValueError("b0_mask must be a boolean array.")
+
+    if not isinstance(dwi, np.ndarray) or dwi.ndim != 4:
+        raise ValueError("dwi must be a 4D numpy array.")
+
+    if dwi.shape[-1] != b0_mask.shape[0]:
+        raise ValueError("The last dimension of dwi must match the length of b0_mask.")
+
+    strategy = strategy.lower()
+    if strategy not in ["first", "all", "mean"]:
+        raise ValueError(
+            "Invalid strategy: {}. Valid strategies are: " "first, all, mean.".format(
+                strategy
+            )
+        )
+
+    if group_contiguous_b0:
+        b0_indices = np.where(b0_mask)[0]
+        groups = np.split(b0_indices, np.where(np.diff(b0_indices) > 1)[0] + 1)
+        grouped_b0 = []
+        for group in groups:
+            group_data = dwi[..., group - b0_indices[0]]
+            if strategy == "first":
+                grouped_b0.append(group_data[..., 0])
+            elif strategy == "mean":
+                grouped_b0.append(np.mean(group_data, axis=-1))
+            else:  # "all"
+                grouped_b0.append(group_data)
+        return np.stack(grouped_b0, axis=-1)
+
+    b0_data = dwi[..., b0_mask]
+    if strategy == "first":
+        return b0_data[..., 0]
+    elif strategy == "mean":
+        return np.mean(b0_data, axis=-1)
+
+    return b0_data
+
+
+def extract_dwi_shell(dwi, gtab, bvals_to_extract, *, tol=20, group_shells=True):
+    """Extract diffusion-weighted imaging (DWI) volumes based on b-value shells.
+
+    Parameters
+    ----------
+    dwi : ndarray
+        4D diffusion-weighted imaging data (X, Y, Z, N), where N is the number of
+        volumes.
+    gtab : GradientTable
+        The gradient table associated with the DWI dataset.
+    bvals_to_extract : list of int
+        List of b-values to extract. If None, all b-values are included.
+    tol : int, optional
+        Tolerance range for b-value selection. A value of 20 means volumes with b-values
+        within ±20 units of the specified b-values will be extracted.
+    group_shells : bool, optional
```

Comment:
Provide the default value details.

Reply:
Please don't. X-ref: https://github.com/dipy/dipy/pull/3444#discussion_r1966687825.